### DISCUSSION TITLE
[2.0] Type hints of execute() results incorrect in PyCharm

### DISCUSSION THREAD

**Discussion by Phanabani** [sqlalchemy/sqlalchemy#9092] [Usage Questions] [answered]
Python 3.10.9
SQLAlchemy 2.0.0rc2
PyCharm 2022.3.1

Hi there!

Apologies in advance if this is a well-discussed issue. I tried searching in many places but couldn't find any answers.

I'm trying to get typing to work for `execute()` results. I followed the [ORM Declarative Models](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#orm-declarative-models) guide using `Mapped` and `mapped_column`, and I saw [this paragraph](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#typing-is-supported-from-step-3-onwards) at the end:

> With the above examples, any example from “step 3” on forward will include that the attributes of the model are typed and will populate through to [select()](https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.select), [Query](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query), and [Row](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Row) objects:

Unfortunately, in PyCharm, the type info is not what I expected. Here's a test file I used (with my problem noted in the comments at the end):

```py
import sqlalchemy as sa
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, sessionmaker


class Base(DeclarativeBase):
    pass


class User(Base):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(sa.Integer, primary_key=True)
    name: Mapped[str]


engine = sa.create_engine("sqlite://")
my_sessionmaker = sessionmaker(engine)
Base.metadata.create_all(engine)

with my_sessionmaker.begin() as session:
    session.add(User(id=123, name="Phanabani"))
    result = session.execute(sa.select(User.name).where(User.id == 123))
    name = result.scalars().one()
    # ^ I press Ctrl+Shift+P or View > Type Info and get the following type for
    # `name`: `Row | RowMapping | Any`, when I instead expect `str`
    assert isinstance(name, str)
    assert name == "Phanabani"
```

Have I done anything wrong, or is this a bug in PyCharm? I also found [this similar discussion](https://github.com/sqlalchemy/sqlalchemy/discussions/8741), but wasn't sure if we were experiencing the same problem and if I should report this to JetBrains.

**Comment by zzzeek** [comment]
your pycharm installation likely has obsolete SQLAlchemy stubs in use.  

Add a reveal_type(), then run mypy:

```py
with my_sessionmaker.begin() as session:
    session.add(User(id=123, name="Phanabani"))
    result = session.execute(sa.select(User.name).where(User.id == 123))
    name = result.scalars().one()

    if TYPE_CHECKING:
        reveal_type(name)

    assert isinstance(name, str)
    assert name == "Phanabani"
```

mypy output:

```
$ mypy test3.py 
test3.py:27: note: Revealed type is "builtins.str"
Success: no issues found in 1 source file

```

it also comes out correctly in pylance / vscode for me (noting that I ensure no obsolete stubs are installed in vscode's private directories)

**Reply by Phanabani** [reply]
Hey thanks so much for the quick response (and the amazing library of course)!

I tried mypy and got the same output as you, so it's clearly a problem with PyCharm. I completely nuked it, uninstalling it and everything related to JetBrains, and after reinstalling the type is still `Row | RowMapping | Any` 😠. I think I'll file an issue with JetBrains then.

**Reply by zzzeek** [reply]
it probably is using sqlalchemy-stubs from typeshed.   you'd have to remove those manually for now.   we cant get typeshed to update until sqlalchemy 2.0 final is released.

**Comment by Phanabani** [comment, marked as answer]
After investigating more, I've realized the issue is that PyCharm does not handle function overloads properly (see issue [PY-42137](https://youtrack.jetbrains.com/issue/PY-42137/PyCharm-type-hinting-doesnt-work-well-with-overload-decorator)). Very frustrating, but that's just how things go sometimes! At least I now know I can use mypy, and might investigate other options.

![image](https://user-images.githubusercontent.com/26840888/212190621-4927ae27-6d6a-4dcb-b913-791822c4b6a3.png)
